include/zephyr/devicetree.h

```diff
@@ -17,9 +17,26 @@
 #define DEVICETREE_H
 
 #include <devicetree_generated.h>
-
 #include <zephyr/sys/util.h>
 
+#ifndef _ASMLANGUAGE
+#include <stdint.h>
+#endif
+
+#ifdef __GNUC__
+/*
+ * This macro must produce constant integer expressions, which can't
+ * be done in the preprocessor (casts aren't allowed).  Defer to the
+ * GCC internal functions where they're available.
+ *
+ * native_posix platform uses it's own version of stdint.h, where
+ * this might not be handled, so we explicitly redefine UINT64_C
+ * macro here
+ */
+#undef UINT64_C
+#define UINT64_C(_v) __UINT64_C(_v)
+#endif
```

Comment:
this is starting to become quite hacky

Reply:
I also don't like this. 
This problem appeared when I tried to write a test on device tree, where the native_posix platform is used. In native (not Zephyr's) version of stdint.h the UINT64_C macro is defined the following way:

```
define UINT64_C(c)   c ## ULL
```

While in Zephyr we have in lib/libc/minimal/include/stdint.h:

```
#ifdef __GNUC__
/* These macros must produce constant integer expressions, which can't
 * be done in the preprocessor (casts aren't allowed).  Defer to the
 * GCC internal functions where they're available.
 */
#define INT8_C(_v) __INT8_C(_v)
#define INT16_C(_v) __INT16_C(_v)
#define INT32_C(_v) __INT32_C(_v)
#define INT64_C(_v) __INT64_C(_v)
#define INTMAX_C(_v) __INTMAX_C(_v)

#define UINT8_C(_v) __UINT8_C(_v)
#define UINT16_C(_v) __UINT16_C(_v)
#define UINT32_C(_v) __UINT32_C(_v)
#define UINT64_C(_v) __UINT64_C(_v)
#define UINTMAX_C(_v) __UINTMAX_C(_v)
#endif /* __GNUC__ */
```

So what options do we have here? Can you suggest something?

Reply:
@mbolivar-nordic , @carlocaione , what do you guys think?

Reply:
@carlocaione , @gmarull , @mbolivar-nordic , guys, do you have any ideas? What else could we do here?

Reply:
Since DT API is used from Linker, you could have something like:

```c
/* include if !_LINKER */
#ifndef _LINKER
#include <stdint.h>
#endif

/* typed macro, document it is equivalent to DT_REG_ADDR in linker context */
#define DT_REG_ADDR_U64(node_id) DT_UINT64_C(DT_REG_ADDR(node_id))

/* ...at the bottom, with helpers */
#ifndef _LINKER
#define DT_UINT64_C(_v) UINT64_C(_v)
#else
#define DT_UINT64_C(_v)
#endif
```

(or add

Reply:
Hi @gmarull , 
Hmmm, that's probably could solve our issue with generic solution. We probably may have add such DT_U64_C for original macros and avoid the usage of specific _U64 versions altogether. 

What do you think?

Reply:
> Hi @gmarull , Hmmm, that's probably could solve our issue with generic solution. We probably may have add such DT_U64_C for original macros and avoid the usage of specific _U64 versions altogether.
> 
> What do you think?

No, because register addresses are not always 64-bit, see e.g. I2C bus addresses.

Reply:
But adding ULL to, say, 8-bit integer literal value will not hurt, and the user remains unaware of the form he receives the address value as far as it is a correct value. From the user's point of view he will use the same
familiar DT_REG_ADDR() macro without caring that under the hood it modifies the address constant in a harmless way.

P.S.: I'm not insisting on this approach, I'm just trying to ensure that we are on the same page. If you want to have explicit _U64 versions of DT_ macros - that's OK, let's have it, even it is slightly ambiguous in relation to 32-bit addresses (in the same way you pointed out with your I2C bus example)

Reply:
Hi @gmarull , @carlocaione , @mbolivar-nordic , 
Guys, let me kindly summarize the remaining questions:

1. What should we do with the redefinition of UINT64_C for native_posix platform in device_tree.h? Shouldn't the native_posix platform use the same stdint.h from lib/libc/minimal/include/stdint.h as the rest of the Zephyr platforms? Currently it looks somehow ugly:

```
#ifdef __GNUC__
/*
 * This macro must produce constant integer expressions, which can't
 * be done in the preprocessor (casts aren't allowed).  Defer to the
 * GCC internal functions where they're available.
 *
 * native_posix platform uses it's own version of stdint.h, where
 * this might not be handled, so we explicitly redefine UINT64_C
 * macro here
 */
#undef UINT64_C
#define UINT64_C(_v) __UINT64_C(_v)
#endif
```

2. @gmarull suggested a nice way to detect linker context, which potentially allows us to have a universal solution for address getters instead of _U64 variants:

```
/* include if !_LINKER */
#ifndef _LINKER
#include <stdint.h>
#endif

#define DT_REG_ADDR(node_id) DT_U64_C(DT_REG_ADDR(node_id))

/* ...at the bottom, with helpers */
#ifndef _LINKER
#define DT_U64_C(_v) UINT64_C(_v)
#else
#define DT_U64_C(_v)
#endif
```

Should we try this way, or should we stay with distinct explicit _U64 variants?
Thanks a lot!

Reply:
go for @gmarull solution, we can comment on the actual implementation.

Reply:
Hi Guys, unfortunately, it looks like the universal solution will not work, since we have macros like:

```
#define DT_PHANDLE_BY_IDX(node_id, prop, idx) \
	DT_CAT6(node_id, _P_, prop, _IDX_, idx, _PH)
#define DT_GPIO_CTLR_BY_IDX(node_id, gpio_pha, idx) \
	DT_PHANDLE_BY_IDX(node_id, gpio_pha, idx)
#define DT_GPIO_CTLR_BY_IDX(DT_BUS(spi_dev), cs_gpios, DT_REG_ADDR(spi_dev))
```
You may see that the value of DT_REG_ADDR(spi_dev) which potentially has ULL at the end will be concatenated with "_PH" which will give some awkward names like DT_N_S_test_S_spi_33334444_P_cs_gpios_IDX_0ULL_PH_ORD

So let's stay with distinct _U64 vesrsions

Reply:
@carlocaione , @gmarull , @mbolivar-nordic , guys please review